src/Fusion/Plugin.hs

```diff
@@ -96,16 +99,198 @@ import Fusion.Plugin.Types
 -- fusion in streamly has become much more predictable which has been verified
 -- by inspecting the core generated by GHC and by inspection testing for the
 -- presence of the stream state constructors.
+
+-------------------------------------------------------------------------------
+-- Set always INLINE on a binder
+-------------------------------------------------------------------------------
+
+unfoldCompulsory :: Arity -> Unfolding -> Unfolding
+unfoldCompulsory arity cuf@CoreUnfolding{} =
+    cuf {uf_src=InlineStable, uf_guidance = UnfWhen arity True True}
+unfoldCompulsory _ x = x -- NoUnfolding
+
+-- Sets the inline pragma on a bndr, and forgets the unfolding.
+setAlwaysInlineOnBndr :: CoreBndr -> CoreBndr
+setAlwaysInlineOnBndr n =
+    let info =
+            case zapUsageInfo $ idInfo n of
+                Just i -> i
+                Nothing ->
+                    error "The impossible happened!! Or GHC changed their api."
+        unf = unfoldingInfo info
+        info' =
+            setUnfoldingInfo
+                (setInlinePragInfo info alwaysInlinePragma)
+                (unfoldCompulsory (arityInfo info) unf)
+     in lazySetIdInfo n info'
+
+--TODO: Replace self-recursive definitions with a loop breaker.
+setInlineOnBndrs :: [CoreBndr] -> CoreExpr -> CoreExpr
+setInlineOnBndrs bndrs = everywhere $ mkT go
+  where
+    go :: CoreExpr -> CoreExpr
+    go (Let (NonRec nn e) expr1)
+        | any (nn ==) bndrs = Let (NonRec (setAlwaysInlineOnBndr nn) e) expr1
+    go x = x
+
+-------------------------------------------------------------------------------
+-- Inspect case alternatives for interesting constructor matches
+-------------------------------------------------------------------------------
+
+-- Checks whether a case alternative contains a type with the
+-- annotation.  Only checks the first typed element in the list, so
+-- only pass alternatives from one case expression.
+altsContainsAnn :: UniqFM [Fuse] -> [Alt CoreBndr] -> Maybe (Alt CoreBndr)
+altsContainsAnn _ [] = Nothing
+altsContainsAnn anns (bndr@(DataAlt dcon, _, _):_) =
+    case lookupUFM anns (getUnique $ dataConTyCon dcon) of
```

Comment:
The result doesn't contain the default alternative, as it is often the first alternative in a list of case statements

Reply:
Yes, does that create an issue here?

Reply:
Yes, a lot of times we actually hit the Default case and not printing that alternative would be misleading.

Reply:
The idea is to show just one of the cases from the `case` statement so that programmer can know which type is involved and where could it be - whichever case it is, it will be the constructor of the same type.